ngraph/core/reference/include/ngraph/runtime/reference/binary_convolution.hpp

```diff
@@ -0,0 +1,227 @@
+                                                 const Shape& f_shape,
+                                                 const Strides& strides,
+                                                 const Strides& dilations,
+                                                 const CoordinateDiff& pads_begin,
+                                                 const CoordinateDiff& pads_end)
+            {
+                // this implementation supports 1D, 2D and 3D convolutions
+                NGRAPH_CHECK(in_shape.size() >= 3 && in_shape.size() <= 5,
+                             "Unsupported input rank: ",
+                             in_shape);
+
+                NGRAPH_CHECK(in_shape.size() == f_shape.size(),
+                             "Incompatible input ranks: ",
+                             in_shape.size(),
+                             " and ",
+                             f_shape.size());
+
+                const auto spatial_dims = in_shape.size() - 2;
+                NGRAPH_CHECK(strides.size() == spatial_dims,
+                             "Strides not definied for all and only spatial dimensions");
+
+                NGRAPH_CHECK(dilations.size() == spatial_dims,
+                             "Dilations not defined for all and only spatial dimensions");
+
+                NGRAPH_CHECK((pads_begin.size() == pads_end.size()) &&
+                                 (pads_begin.size() == spatial_dims),
+                             "Pads not defined for all and only spatial dimensions");
+            }
+
+            template <typename T_IN, typename T_F>
+            void binary_convolution(const T_IN* in,
+                                    const T_F* f,
+                                    T_IN* out,
+                                    const Shape& in_shape,
+                                    const Shape& f_shape,
+                                    const Shape& out_shape,
+                                    const Strides& strides,
+                                    const Strides& dilations,
+                                    const CoordinateDiff& pads_begin,
+                                    const CoordinateDiff& pads_end,
+                                    const float pad_value)
+            {
+                validate_convolution_parameters(
+                    in_shape, f_shape, strides, dilations, pads_begin, pads_end);
+
+                // here we are converting all param types to int's to avoid arithmetic issues
+                // (e.g signed + unsigned) in indexes calculation later
+                ConvolutionParams params{strides, dilations, pads_begin, pads_end};
+
+                // here we are extending spatial dimensions to 3D, because we are going to use 3D
+                // convolution implementation to convolve also in 1D & 2D case
+                Shape input_shape{in_shape};
+                Shape filters_shape{f_shape};
+                if (in_shape.size() < 5)
```

Comment:
What if `in_shape.size() > 5`? Should we add a check?

Reply:
Sure. This check is done [here](https://github.com/openvinotoolkit/openvino/pull/4278/files#diff-59e0f3820bc7a8c78344cc7f4cd11d4e1935bf9e2a1415c46a68e418c696a1ffR150) in helper function [validate_convolution_parameters](https://github.com/openvinotoolkit/openvino/pull/4278/files#diff-59e0f3820bc7a8c78344cc7f4cd11d4e1935bf9e2a1415c46a68e418c696a1ffR185)